gabriel-portfolio/portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -13,20 +13,88 @@
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
 import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Collections;
+import com.google.gson.Gson;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
-/** Servlet that returns some example content. TODO: modify this file to handle comments data */
-// Did not remove this because it will be used in week 3
-// @WebServlet("/data")
-// public class DataServlet extends HttpServlet {
-//   @Override
-//   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-//     response.setContentType("text/html;");
-//     response.getWriter().println("<h1>Hello world!</h1>");
-//   }
-// }
+
+@WebServlet("/data")
+public class DataServlet extends HttpServlet {
+  
+  final String commentEntity = "Comment";
+  final String textProperty = "text";
+  final String timestampProperty = "timestamp";
+
+  static Gson gson;
+
+  @Override
+  public void init() {
+    gson = new Gson();
+  }
+  
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+    Query query = new Query(commentEntity).addSort(timestampProperty, SortDirection.DESCENDING);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    // loop through the query to get the text from each comment
+    List<String> comments = new ArrayList<String>();
+    int commentCounter = 0;
+    
+    loopThroughComments: 
+    for(Entity e : results.asIterable()) {
+        String text = (String) e.getProperty(textProperty);
+        if(!(text == null)) {
+          comments.add(text);
+          commentCounter++;
+        }
+    }
+    
+    // convert string to json and output
+    String json = gson.toJson(comments);
+    response.setContentType("application/json;");
+    response.getWriter().println(json);
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    
+    String text = getParameter(request, "text-input", "");
+    long timestamp = System.currentTimeMillis();
+    
+    Entity comment = new Entity(commentEntity);
+    comment.setProperty(textProperty, text);
+    comment.setProperty(timestampProperty, timestamp);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.put(comment);
+
+    response.sendRedirect("/index.html");
+  }
+
+  /*
+    Returns a certain parameter from a post requests
+    if it doesn't exist, it returns the a default value 
+  */
+  private String getParameter(HttpServletRequest request, String name, String defaultValue) {
+    String value = request.getParameter(name);
+    if (value == null) {
```

Comment:
You can replace this with the ternary expression to be cleaner:
return value == null ? defaultValue : value;